indexer/categories_holder.cpp

```diff
@@ -138,68 +246,33 @@ void CategoriesHolder::LoadFromStream(istream & s)
 
     strings::SimpleTokenizer iter(line, state == EParseTypes ? "|" : ":|");
 
-    switch (state)
-    {
-    case EParseTypes:
+    if (state == EParseTypes)
     {
       AddCategory(cat, types);
       currentGroups.clear();
 
       while (iter)
       {
-        // Check if category is a group reference.
-        if ((*iter)[0] == '@')
-        {
-          CHECK((currentGroups.empty() || !types.empty()),
-                ("Two groups in a group definition at line", lineNumber));
-          currentGroups.push_back(*iter);
-        }
-        else
-        {
-          // Split category to subcategories for classificator.
-          vector<string> v;
-          strings::Tokenize(*iter, "-", MakeBackInsertFunctor(v));
-
-          // Get classificator type.
-          uint32_t const type = c.GetTypeByPathSafe(v);
-          if (type != 0)
-            types.push_back(type);
-          else
-            LOG(LWARNING, ("Invalid type:", v, "at line:", lineNumber));
-        }
-
+        ProcessCategory(*iter, currentGroups, types);
         ++iter;
       }
 
       if (!types.empty() || currentGroups.size() == 1)
         state = EParseLanguages;
     }
-    break;
-
-    case EParseLanguages:
+    else if (state == EParseLanguages)
     {
       if (!iter)
       {
-        // If the category groups are specified, add translations from them.
-
-        ///@todo According to the current logic, categories.txt should have
-        /// the blank string at the end of file.
-        if (!types.empty())
-        {
-          for (string const & group : currentGroups)
-          {
-            auto it = m_groupTranslations.find(group);
-            if (it == m_groupTranslations.end())
-              continue;
-            for (auto const & synonym : it->second)
-              cat.m_synonyms.push_back(synonym);
-          }
-        }
-
         state = EParseTypes;
         continue;
       }
 
+      // Add translations into synonyms first, it will allow to override
+      // translations for UI by concrete category translation.
+      if (!types.empty() && cat.m_synonyms.empty())
```

Comment:
Please, move this check to the place when state switches from type parsing to languages parsing.